0000-settlement-engines/0000-settlement-engines.md

```diff
@@ -0,0 +1,395 @@
+
+#### Request
+
+```http
+POST /accounts/:id/messages HTTP/1.1
+Accept: application/octet-stream
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of message&gt;_
+
+#### Response
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of response&gt;_
+
+## Accounting System HTTP API
+
+_Settlement engine &rarr; connector_
+
+HTTP/2 is RECOMMENDED for performance reasons, although HTTP/1.1 MAY also be used. Implementations SHOULD support HTTP version negotiation via Application Protocol Negotiation (ALPN).
+
+### Credit incoming settlement
+
+[Account for an incoming settlement](#account-for-incoming-settlements). The settlement engine sends this request as it receives incoming settlements in in order to ensure [settlement symmetry](#settlement-symmetry). If the accounting system responds that it only credited a partial amount, the settlement engine [tracks the uncredited amounts](#track-uncredited-incoming-settlements).
+
+#### Request
+
+```http
+POST /accounts/:id/settlements HTTP/1.1
+Accept: application/json
+Content-Type: application/json
+Idempotency-Key: <key>
+```
+
+> **[`Quantity`](#quantity-json-type)** to be credited to the account as an incoming settlement
+
+#### Response
+
+```http
+HTTP/1.1 201 CREATED
+Content-Type: application/json
+```
+
+> **[`Quantity`](#quantity-json-type)** credited to the account as an incoming settlement, which MUST  be less than or equal to the quantity in the original request
+
+### Send outgoing message
+
+Send the message to the given peer's settlement engine and return its response. The connector handles [proxying the message](#exchanging-messages) through the peer's connector.
+
+#### Request
+
+```http
+POST /accounts/:id/messages HTTP/1.1
+Accept: application/octet-stream
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of message&gt;_
+
+#### Response
+
+```http
+HTTP/1.1 200 OK
+Content-Type: application/octet-stream
+```
+
+> _&lt;raw bytes of response&gt;_
+
+## Idempotence
+
+Idempotent requests ensure each side effect only happens once, even though the same request may be called multiple times. For example, if a request to perform a settlement fails due to a network connection error, [idempotence](https://en.wikipedia.org/wiki/Idempotence) prevents a client from accidentally triggering multiple settlements when it retries the request.
+
+### Performing idempotent requests
+
+Requests to settle or requests to credit incoming settlements MUST include an idempotency key, or globally unique string, within an `Idempotency-Key: <key>` header. To avoid collisions, this key MUST be derived from a cryptographically secure source of randomness.
+
+### Recommended retry behavior
+
+If the client receives no response, it MUST retry the request with the same idempotency key.
```

Comment:
```suggestion
If the client receives no response for an idempotent request, the client SHOULD retry the request with the same idempotency key.
```

Reply:
This is a MUST. It's unsafe if it doesn't retry; the spec has been updated to mandate that clients retry indefinitely.